### PR TITLE
Qt: Network Watch tool

### PR DESCRIPTION
**Pull request by luke-jr**
Simple realtime log of p2p network activity (blocks and transactions only)

- Doesn't begin logging until opened; limited to 0x400 entries (outputs)
- Automatically scrolls if left at the bottom of the log; maintains position if left elsewhere
- Memory-efficient circular buffer; CTransaction references become weak after they're 0x200 entries back in the log
- Search function that selects all matching log entries, including ongoing

### CODE REVIEW

File: src/qt/netwatch.cpp [hunk 1]
```diff
@@ -0,0 +1,788 @@
...
+        getLogEntryRow(rows_used) = le;
+        --logskip;
+    } else {
+        // Haven't filled up yet, so just push_back
+        assert(!logpos);
+        log.push_back(le);
+    }
+    ++rows_used;
+    if (rows_used > max_nonweak_txouts) {
+        LogEntry& le = getLogEntryRow(rows_used - max_nonweak_txouts - 1);
```
Comment (paveljanik) [inline src/qt/netwatch.cpp]: `le` is the same name as the first argument here, it will bring shadow warning.

**Comment by paveljanik** [comment]
This is how it looks like here on testnet:

screen shot 2017-02-24 at 08 22 39

**Comment by jonasschnelli** [comment]
Thanks. Will review.
I just played a bit with it and has massive locking issues on mainnet during catch-up of 2-3 weeks.

**Comment by jonasschnelli** [comment]
Running this PR (built over gitian: https://bitcoin.jonasschnelli.ch/build/54) makes Bitcoin-Qt and also the rest of my apps almost unusable. Had to force kill the process.

File: src/qt/netwatch.cpp [hunk 2]
```diff
@@ -0,0 +1,798 @@
...
+        if (!new_matches.isEmpty()) {
+            Q_EMIT moreSearchResults(new_matches);
+        }
+    }
+}
+
+void NetWatchLogModel::LogBlock(const CBlockIndex* pblockindex)
+{
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pblockindex, Params().GetConsensus())) {
```
Comment (jonasschnelli) [inline src/qt/netwatch.cpp]: IMO reading each block will cause a massive slow down during IBD / catchup.

File: src/qt/netwatch.cpp [hunk 3]
```diff
@@ -0,0 +1,798 @@
...
+        // Indicate error somehow?
+        return;
+    }
+    assert(block.vtx.size());
```
Comment (jonasschnelli) [inline src/qt/netwatch.cpp]: What we probably should do is adding each blocks size and vtx.size() to CBlockIndex* (would require to alter the block index, migration, etc. yes).

File: src/qt/netwatch.cpp (hunk 2 above)
Comment (luke-jr) [inline src/qt/netwatch.cpp]: The OS should have the data cached already?

File: src/qt/netwatch.cpp (hunk 3 above)
Comment (luke-jr) [inline src/qt/netwatch.cpp]: That wouldn't change anything here...?

**Comment by laanwj** [comment]
Concept ACK, I like having this, though we have to be sure that there is *no* performance impact when the monitor is not running, and as little as possible when it is.

**Comment by luke-jr** [comment]
The code is (already) disabled until the first time it is opened.

**Comment by luke-jr** [comment]
Rebased

**Comment by luke-jr** [comment]
Oops, had rebased onto 0.17 instead of master. Rebased for real this time.

File: src/qt/netwatch.cpp [hunk 4]
```diff
@@ -0,0 +1,838 @@
...
+}
+
+void NetWatchLogModel::setClientModel(ClientModel *model)
+{
+    if (clientModel) {
+        disconnect(clientModel->getOptionsModel(), SIGNAL(displayUnitChanged(int)), this, SLOT(updateDisplayUnit()));
+    }
+    clientModel = model;
+    if (model) {
+        connect(model->getOptionsModel(), SIGNAL(displayUnitChanged(int)), this, SLOT(updateDisplayUnit()));
```
Comment (MarcoFalke) [inline src/qt/netwatch.cpp]: Might want to use the new connect syntax in this file?

File: src/qt/bitcoingui.cpp [hunk 5]
```diff
@@ -661,6 +676,18 @@ void BitcoinGUI::aboutClicked()
...
+void BitcoinGUI::showNetWatch()
+{
+    if (!NetWatch) {
+        NetWatch = new GuiNetWatch(platformStyle, netStyle);
+        NetWatch->setClientModel(clientModel);
+    }
+    NetWatch->showNormal();
+    NetWatch->show();
+    NetWatch->raise();
+    NetWatch->activateWindow();
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Could rebase with #14123 to replace these calls with `GUIUtil::bringToFront`.

File: src/qt/bitcoingui.cpp [hunk 6]
```diff
@@ -73,7 +74,8 @@ const std::string BitcoinGUI::DEFAULT_UIPLATFORM =
 BitcoinGUI::BitcoinGUI(interfaces::Node& node, const PlatformStyle *_platformStyle, const NetworkStyle *networkStyle, QWidget *parent) :
     QMainWindow(parent),
     m_node(node),
-    platformStyle(_platformStyle)
+    platformStyle(_platformStyle),
+    netStyle(networkStyle)
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Remove.

File: src/qt/netwatch.h [hunk 7]
```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2017 The Bitcoin Core developers
```
Comment (promag) [inline src/qt/netwatch.h]: nit 2018.

File: src/qt/netwatch.h [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+class QLineEdit;
+class QTableView;
+QT_END_NAMESPACE
+
+class CBlock;
+class CBlockIndex;
+class ClientModel;
+class LogEntry;
+class PlatformStyle;
+class NetworkStyle;
```
Comment (promag) [inline src/qt/netwatch.h]: nit, sort.

File: src/qt/netwatch.h [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+static const int nLongestAddress = 35;
```
Comment (promag) [inline src/qt/netwatch.h]: move to .cpp? nit, follow code convention.

File: src/qt/bitcoingui.cpp [hunk 10]
```diff
@@ -397,6 +404,7 @@ void BitcoinGUI::createMenuBar()
     settings->addAction(optionsAction);
 
     QMenu *help = appMenuBar->addMenu(tr("&Help"));
+    help->addAction(NetWatchAction);
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: FYI in #14573 I'm proposing a Window menu.

File: src/qt/bitcoingui.h [hunk 11]
```diff
@@ -127,6 +128,7 @@ class BitcoinGUI : public QMainWindow
     QAction* backupWalletAction = nullptr;
     QAction* changePassphraseAction = nullptr;
     QAction* aboutQtAction = nullptr;
+    QAction* NetWatchAction = nullptr;
```
Comment (promag) [inline src/qt/bitcoingui.h]: nit, `m_show_netwatch_action`.

File: src/qt/netwatch.cpp [hunk 12]
```diff
@@ -0,0 +1,838 @@
...
+        }
+        if (count == txout_index) {
+            return &txout;
+        }
+        ++count;
+    }
+    return NULL;
+}
+
+}
```
Comment (promag) [inline src/qt/netwatch.cpp]: nit, `} // namespace`

File: src/qt/netwatch.cpp [hunk 13]
```diff
@@ -0,0 +1,838 @@
...
+
+const QString LogEntry::LogEntryTypeAbbreviation(const LogEntryType LET)
```
Comment (promag) [inline src/qt/netwatch.cpp]: nit, s/LET/log_entry_type.

File: src/qt/netwatch.cpp [hunk 14]
```diff
@@ -0,0 +1,838 @@
...
+
+bool IsDatacarrier(const CTxOut& txout)
+{
+    return (txout.scriptPubKey[0] == OP_RETURN && txout.nValue == 0);
+}
+
+size_t CountNonDatacarrierOutputs(const CTransactionRef& tx)
+{
+    size_t count = 0;
+    for (auto& txout : tx->vout) {
```
Comment (promag) [inline src/qt/netwatch.cpp]: const?

File: src/qt/netwatch.cpp [hunk 15]
```diff
@@ -0,0 +1,838 @@
...
+        if (IsDatacarrier(txout)) {
```
Comment (promag) [inline src/qt/netwatch.cpp]: nit, one line `if (...) continue`;

File: src/qt/netwatch.cpp [hunk 16]
```diff
@@ -0,0 +1,838 @@
...
+        ++count;
+    }
+    return count;
+}
+
+const CTxOut* GetNonDatacarrierOutput(const CTransactionRef& tx, const size_t txout_index)
+{
+    size_t count = 0;
+    for (auto& txout : tx->vout) {
+        if (IsDatacarrier(txout)) {
```
Comment (promag) [inline src/qt/netwatch.cpp]: Same as above.

File: src/qt/netwatch.cpp [hunk 17]
```diff
@@ -0,0 +1,838 @@
...
+            continue;
+        }
+        if (count == txout_index) {
```
Comment (promag) [inline src/qt/netwatch.cpp]: Same as above.

File: src/qt/netwatch.cpp [hunk 18]
```diff
@@ -0,0 +1,838 @@
...
+            return &txout;
+        }
+        ++count;
+    }
+    return NULL;
```
Comment (promag) [inline src/qt/netwatch.cpp]: `nullptr`.

File: src/qt/netwatch.cpp [hunk 19]
```diff
@@ -0,0 +1,838 @@
+// Copyright (c) 2017 The Bitcoin Core developers
```
Comment (promag) [inline src/qt/netwatch.cpp]: nit, 2018

**Review by promag — COMMENTED** [review]
Concept ACK.

Code style is a bit outdated.

File: src/qt/netwatch.cpp [hunk 20]
```diff
@@ -0,0 +1,838 @@
...
+    }
+    CTransactionRef * const ptx_old = get<CTransactionRef>();
+    CTransactionRef tx = *ptx_old;  // save a copy
+    ptx_old->~CTransactionRef();
+    CTransactionWeakref * const ptx_new = get<CTransactionWeakref>();
+    new (ptx_new) CTransactionWeakref(tx);
+    *((meta_t*)data) |= (3 << 30);
+}
+
+class NetWatchValidationInterface final : public CValidationInterface {
```
Comment (promag) [inline src/qt/netwatch.cpp]: Should use/extend `interfaces::Node` instead?

File: src/qt/bitcoingui.cpp (hunk 6 above)
Comment (luke-jr) [reply src/qt/bitcoingui.cpp]: It's needed...

File: src/qt/netwatch.h (hunk 7 above)
Comment (luke-jr) [reply src/qt/netwatch.h]: 2018 is wrong. This code hasn't been substantially changed since 2017.

File: src/qt/netwatch.cpp (hunk 17 above)
Comment (luke-jr) [reply src/qt/netwatch.cpp]: IMO this one is clearer split out.

File: src/qt/netwatch.h [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+
+public:
+    enum LogEntryType {
+        LET_BLOCK,
+        LET_TX,
+    };
+    static const QString LogEntryTypeAbbreviation(LogEntryType);
+
+    LogEntry(const LogEntry&);
+    LogEntry() : data(NULL) { }
```
Comment (practicalswift) [inline src/qt/netwatch.h]: Use `nullptr` instead of `NULL`. Applies throughout this PR :-)

File: src/qt/bitcoingui.cpp (hunk 6 above)
Comment (promag) [reply src/qt/bitcoingui.cpp]: Sorry, forgot to delete this comment while reviewing..

File: src/qt/netwatch.cpp [hunk 22]
```diff
@@ -0,0 +1,829 @@
...
+
+#include <qt/netwatch.h>
+
+#include <qt/bitcoinunits.h>
+#include <qt/clientmodel.h>
+#include <qt/guiconstants.h>
+#include <qt/guiutil.h>
+#include <qt/optionsmodel.h>
+#include <qt/platformstyle.h>
+#include <qt/networkstyle.h>
```
Comment (practicalswift) [inline src/qt/netwatch.cpp]: Nit: Sort includes :-)

File: src/qt/netwatch.cpp [hunk 23]
```diff
@@ -0,0 +1,829 @@
...
+    new (ptx_new) CTransactionWeakref(tx);
+    *((meta_t*)data) |= (3 << 30);
+}
+
+class NetWatchValidationInterface final : public CValidationInterface {
+private:
+    NetWatchLogModel& model;
+
+public:
+    NetWatchValidationInterface(NetWatchLogModel& model_in) : model(model_in) {}
```
Comment (practicalswift) [inline src/qt/netwatch.cpp]: Should be `explicit`? :-)

File: src/qt/netwatch.cpp [hunk 24]
```diff
@@ -0,0 +1,829 @@
...
+        case NWLMH_VALUE: {
+            if (blockindex.nTx == 0 || !(blockindex.nStatus & BLOCK_HAVE_DATA)) {
+                return QVariant();
+            }
+            CBlock block;
+            if (!ReadBlockFromDisk(block, &blockindex, Params().GetConsensus())) {
+                // Indicate error somehow?
+                return QVariant();
+            }
+            assert(block.vtx.size());
```
Comment (practicalswift) [inline src/qt/netwatch.cpp]: Nit: `!block.vtx.empty()` communicates intention more clearly than `block.vtx.size()`?

File: src/qt/netwatch.cpp [hunk 25]
```diff
@@ -0,0 +1,829 @@
...
+    }
+    return false;
+}
+
+void NetWatchLogModel::searchRows(const QString& query, QList<int>& results)
+{
+    const auto currentUnit = clientModel->getOptionsModel()->getDisplayUnit();
+    NetWatchLogSearch *newsearch = new NetWatchLogSearch(query, currentUnit);
+    LOCK(cs);
+    if (currentSearch) {
```
Comment (practicalswift) [inline src/qt/netwatch.cpp]: This `if` statement is unnecessary. Deleting a null pointer is a noop.

File: src/qt/netwatch.h [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+    static const int NWLMHeaderCount = 5;
+    enum Header {
+        NWLMH_TIME,
+        NWLMH_TYPE,
+        NWLMH_ID,
+        NWLMH_ADDR,
+        NWLMH_VALUE,
+    };
+
+    NetWatchLogModel(QWidget *parent);
```
Comment (practicalswift) [inline src/qt/netwatch.h]: Should be made `explicit`?

File: src/qt/netwatch.cpp [hunk 27]
```diff
@@ -0,0 +1,829 @@
...
+    void UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload);
+    void TransactionAddedToMempool(const CTransactionRef &);
+};
+
+void NetWatchValidationInterface::ValidationInterfaceUnregistering()
+{
+    model.OrphanedValidationInterface();
+}
+
+void NetWatchValidationInterface::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload)
```
Comment (practicalswift) [inline src/qt/netwatch.cpp]: Should be marked `override`?

File: src/qt/netwatch.cpp [hunk 28]
```diff
@@ -0,0 +1,829 @@
...
+{
+    if (pindexNew != pindexFork) {
+        model.LogBlock(pindexNew);
+    }
+}
+
+void NetWatchValidationInterface::TransactionAddedToMempool(const CTransactionRef &ptx)
```
Comment (practicalswift) [inline src/qt/netwatch.cpp]: Should be marked `override`?

File: src/qt/netwatch.h [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+        void *p = (void *)intptr_t(sizeof(meta_t));
+        size_t dummy_bufsize = sizeof(T) * 2;
+        p = align(std::alignment_of<T>::value, sizeof(T), p, dummy_bufsize);
+        assert(p);
+        offset = size_t(p);
+        return offset + sizeof(T);
+    }
+
+    // std::align (missing in at least GCC 4.9) substitute from c-plus
+    static inline void *align( std::size_t alignment, std::size_t size, void *&ptr, std::size_t &space ) {
```
Comment (practicalswift) [inline src/qt/netwatch.h]: Remove extra spaces after `(` and before `)` :-)

File: src/qt/netwatch.cpp (hunk 24 above)
Comment (luke-jr) [reply src/qt/netwatch.cpp]: Not sure. The intent is only to make sure `vtx[0]` below is an acceptable dereference.

The actual check for block data is above, with `if (blockindex.nTx == 0 || !(blockindex.nStatus & BLOCK_HAVE_DATA))`

Comment (practicalswift) [reply src/qt/netwatch.cpp]: I see -- makes sense :-)

File: src/qt/netwatch.cpp (hunk 27 above)
Comment (practicalswift) [reply src/qt/netwatch.cpp]: Sounds good. I commented on the wrong line :-)

File: src/qt/netwatch.cpp [hunk 30]
```diff
@@ -0,0 +1,827 @@
...
+void NetWatchLogModel::log_append(const LogEntry& le, size_t& rows_used)
+{
+    AssertLockHeld(cs);
+    if (m_logskip) {
+        // Replace a deleted row
+        getLogEntryRow(rows_used) = le;
+        --m_logskip;
+    } else {
+        // Haven't filled up yet, so just push_back
+        assert(!m_logpos);
```
Comment (ken2812221) [inline src/qt/netwatch.cpp]: I see this assertion on Windows, haven't take a look at the code.

Comment (luke-jr) [reply src/qt/netwatch.cpp]: Any idea how?>

File: src/qt/netwatch.cpp [hunk 31]
```diff
@@ -0,0 +1,827 @@
...
+    const meta_t n = *(meta_t*)m_data;
+    if ((n >> 30) != 2) {
+        return;
+    }
+    CTransactionRef * const ptx_old = get<CTransactionRef>();
+    CTransactionRef tx = *ptx_old;  // save a copy
+    ptx_old->~CTransactionRef();
+    CTransactionWeakref * const ptx_new = get<CTransactionWeakref>();
+    new (ptx_new) CTransactionWeakref(tx);
+    *((meta_t*)m_data) |= (3 << 30);
```
Comment (practicalswift) [inline src/qt/netwatch.cpp]: Shift overflow?

**Comment by laanwj** [comment]
As this PR has been inactive for a long time, I'm closing this (but marking "up for grabs").

**Comment by MarcoFalke** [comment]
Moved to https://github.com/bitcoin-core/gui/pull/444

**Pull request closed without merging**